src/lua/datetime.lua

```diff
@@ -420,6 +492,208 @@ local function datetime_tostring(self)
     return s
 end
 
+local function qtail(s)
+    return #s ~= 0 and (s .. ', ') or ''
+end
+
+--[[
+    Convert to text interval values of different types
+
+    - depending on a values stored there generic interval
+      values may display in following format:
+        +12 secs
+        -23 minutes, 0 seconds
+        +12 hours, 23 minutes, 1 seconds
+        -7 days, -23 hours, -23 minutes, -1 seconds
+    - years will be displayed as
+        +10 years
+    - months will be displayed as:
+         +2 months
+]]
+local function interval_tostring(o)
+    check_interval(o, 'datetime.interval.tostring')
+    local s = ''
+    if o.year ~= 0 then
+        s = qtail(s) .. ('%+d years'):format(o.year)
+    end
+    if o.month ~= 0 then
+        s = qtail(s) .. ('%+d months'):format(o.month)
+    end
+    local ts = o.sec + o.nsec / 1e9
+    if ts == 0 then
+        return #s > 0 and s or '0 seconds'
+    end
+    local abs_ts = ts < 0 and -ts or ts
+    if abs_ts < 60 then
+        s = qtail(s) .. ('%+.9g seconds'):format(ts)
+    elseif abs_ts < (60 * 60) then
+        s = qtail(s) .. ('%+d minutes, %.9g seconds'):
+                        format(o.min, math_fmod(ts, 60))
+    elseif abs_ts < (24 * 60 * 60) then
+        s = qtail(s) .. ('%+d hours, %d minutes, %.9g seconds'):
+                        format(o.hour, math_fmod(o.min, 60), math_fmod(ts, 60))
+    else
+        s = qtail(s) .. ('%+d days, %d hours, %d minutes, %.9g seconds'):
+                        format(o.day, math_fmod(o.hour, 24),
+                               math_fmod(o.min, 60), math_fmod(ts, 60))
+    end
+    return s
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 or nsec >= NANOS_PER_SEC then
+        secs = secs + math_floor(nsec / NANOS_PER_SEC)
+        nsec = nsec % NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_increment_by(self, direction, years, months,
+                                     seconds, nanoseconds)
+    -- operations with intervals should be done using human dates
+    -- not UTC dates, thus we normalize to UTC
+    local dt = local_dt(self)
+    local secs, nsec = local_secs(self), self.nsec
+    local offset = self.tzoffset
+
+    local ym_updated = false
+
+    if years ~= 0 then
+        check_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR, 'years')
```

Comment:
So are you saying that there are still such dates and intervals, that their sum leads to the overflow? If so, then I propose to fix it pessimistically unless we can do a precise check. For instance, to check for year overflow use 366 days, not 365.25. 31 days for months. Currently the patch does not respect the boundaries exactly anyway, but still allows the overflow (if I understood it correctly).
Can you find such a date + interval combination that in your patch it still leads to the overflow?

Reply:
Using 366 days for year, and 31 days for month would _extend_ range of accepted values (which would end up with silent overflow/underflow in c-dt code), so I'd rather use range which slightly _shrink_ accepted range (like we use today). This way there will be no twilight zone when we allow in our Lua code, silently ignore in c-dt and underlow/overflow underlying `dt_t` value.

Reply:
@Gerold103, If I understand it correctly, using 365.25 days / year is enough to ensure there are no overflows.
Gregorian calendar has ~ 365.2425 days / year, which's smaller.
So we're safe.

I'm only concerned with "dt.new{year=5879611, month=5, day=12} + {month=2}"
I feel like the boundary check should pass on this one, resulting in an overflow. But the boundary check doesn't pass, and I don't understand why. (2 * AVERAGE_DAYS_MONTH = 60.875, and adding 60.875 days to {year=5879611, month=5, day=12} is totally ok)

Reply:
> I'm only concerned with "dt.new{year=5879611, month=5, day=12} + {month=2}"
> I feel like the boundary check should pass on this one, resulting in an overflow. But 
> the boundary check doesn't pass.

I don't actually understand what is confusing you here. Looks like it's properly handling this boundary

```
tarantool> date.new{year=5879611, month=5, day=12} + {month=2}
---
- error: '[string "return date.new{year=5879611, month=5, day=12..."]:1: addition
    of 2 months makes date greater than maximum allowed 5879611-07-11'
...

tarantool> date.new{year=5879611, month=5, day=11} + {month=2}
---
- 5879611-07-11T00:00:00Z
...
```

Reply:
> I don't actually understand what is confusing you here. Looks like it's properly handling this boundary

The behaviour is correct. But I don't understand why.

There are 61 days in 2 months between may 12 and jul 12.
But 2 * AVERAGE_DAYS_MONTH is 60.875, which is less than 61.

Your boundary check tries to add 60.875 days to may 12.
This should be perfectly fine, since it results in:
```
tarantool> dt.new{year=5879611, month=5, day=12}:add{day=60.875}
---
- 5879611-07-11T21:00:00Z
...

```

Still, the boundary check fails (which is correct). But I don't understand why it fails.

Reply:
It was properly failing for `:add{month = 2}` because resultant `dt` value was becoming `2147483647.875` which is greater than maximum int allowed (`2147483647`). 

And it was not failing for `:add{days = 60.875}` because there was no check for "seconds" flow, only for "years" and "months". That's oversight. Thanks for asking!

Reply:
I rather differently - there was check in "seconds" flow, but there was redundant `math.floor` called, which was working wrongly on boundary days.

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -720,7 +720,7 @@ local function datetime_increment_by(self, direction, years, months,
     end
 
     secs, self.nsec = normalize_nsec(secs, nsec)
-    check_dt(math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY), direction)
+    check_dt((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY, direction)
     self.epoch = utc_secs(secs, offset)
     return self
 end
```

@sergepetrenko thanks again for paying attention!

Reply:
@tsafin, thanks for figuring this out!